shared/url_generation.py

```diff
@@ -0,0 +1,192 @@
+    :param bucket_name:
+    :param new_offset_in_seconds:
+    :return:
+    """
+
+    log = regular_log.default()
+    connection = Connection.get_by_id(session = session, id = connection_id)
+    if connection is None:
+        msg = f'connection id: {connection_id} not found.'
+        log['error']['connection_id'] = msg
+        logger.error(msg)
+        return blob_object, log
+    member = get_member(session = session)
+    params = {
+        'bucket_name': bucket_name,
+        'path': blob_object.url_signed_blob_path if reference_file is None else reference_file.bl,
+        'expiration_offset': new_offset_in_seconds,
+        'action_type': 'get_pre_signed_url',
+        'event_data': {
+            'request_user': member.user_id
+        }
+    }
+
+    if connection.integration_name not in ALLOWED_CONNECTION_SIGNED_URL_PROVIDERS:
+        msg = f'Unsupported connection provider for URL regeneration {connection.id}:{connection.integration_name}'
+        log['error']['unsupported'] = msg
+        logger.error(msg)
+        return blob_object, log
+
+    connection_strategy = ConnectionStrategy(
+        connection_class = S3Connector,
+        connector_id = connection_id,
+        session = session)
+
+    client, success = connection_strategy.get_connector(connector_id = connection_id)
+    if not success:
+        msg = f'Failed to get connector for connection {connection_id}'
+        log['error']['connector'] = msg
+        logger.error(msg)
+        return blob_object, log
+    signed_url, log = get_url_from_connector(connector = client, params = params, log = log)
+    if regular_log.log_has_error(log):
+        return blob_object, log
+    if regular_log.log_has_error(log):
+        return blob_object, log
+    blob_object.url_signed = signed_url
+
+    # Extra assets (Depending on type)
+    if type(blob_object) == Image and blob_object.url_signed_thumb_blob_path:
+        params['path'] = blob_object.url_signed_thumb_blob_path
+        thumb_signed_url, log = get_url_from_connector(connector = client, params = params, log = log)
+        if regular_log.log_has_error(log):
+            return blob_object, log
+        blob_object.url_signed_thumb = thumb_signed_url
+
+    return blob_object, log
+
+
+def blob_regenerate_url(blob_object: DiffgramBlobObjectType,
+                        session: Session,
+                        connection_id = None,
+                        bucket_name = None,
+                        reference_file: File = None):
+    """
+        Regenerates the signed url of the given blob object.
+    :param blob_object:
+    :param session:
+    :param connection_id:
+    :param bucket_name:
+    :return:
+    """
+    if not blob_object.url_signed_blob_path:
+        return
+    should_regenerate, new_offset_in_seconds = data_tools.determine_if_should_regenerate_url(blob_object, session)
+
+    if should_regenerate is not True: 
+        return
+
+    strategy = determine_url_regenerate_strategy(
```

Comment:
added example pattern https://github.com/diffgram/diffgram/commit/92268c223d9901dfa7432eb0c7104223fb1fba4b

Reply:
Super looks great will add this on the upload step too!